### PR TITLE
Split `Predicate<T, U extends T>` into `Predicate<T>` and `Refinement<T, U extends T>`

### PR DESCRIPTION
**Pull request by kasperisager**
This pull request splits the current `Predicate<T, U extends T>` into two separate types:

1. `Predicate<T>`.
2. `Refinement<T, U extends T>`, living in the new `@siteimprove/alfa-refinement` package.

`Predicate<T>` now always acts as a plain predicate, performing no type narrowing, and `Refinement<T, U extends T>` always acts as a type predicate, narrowing the passed `T` to a `U`. This solves a long-standing issue wherein plain predicates of type `Predicate<T, T>` were, in certain situations, being inferred as type guards, causing the `else` branch of whatever code was using the predicate to narrow the corresponding value to `never`:

```ts
const isFoo: Predicate<string, string> = (value) => value === "foo";

const value: string = "bar";

if (isFoo(value)) {
  // value is `string`
} else {
  // value is `never`
}
```

To circumvent this, the `Predicate.test()` method had to be used, often requiring an explicit type annotation to avoid the incorrect inference, as seen here:

https://github.com/Siteimprove/alfa/blob/f07ab0eb791d7dbdd8a8c6647518cda090f1ca56/packages/alfa-aria/src/feature.ts#L378-L417

By separating these two behaviours, narrowing and non-narrowing, into separate types, we ensure that stuff like this doesn't happen. This separation also paves the way for #410, which will allow us to further clean up code like seen above.

### CODE REVIEW

File: packages/alfa-refinement/test/refinement.spec.ts [hunk 1]
```diff
@@ -0,0 +1,75 @@
+// import { test } from "@siteimprove/alfa-test";
```
Comment (Jym77) [inline packages/alfa-refinement/test/refinement.spec.ts:1]: Tests should be active.

File: packages/alfa-predicate/test/predicate.spec.ts [hunk 2]
```diff
@@ -1,75 +1,75 @@
...
-      | null
-      | undefined
-      | symbol = value;
-
-    t.equal(foo, 123);
-  } else {
-    t.fail();
-  }
-});
+// import { test } from "@siteimprove/alfa-test";
```
Comment (Jym77) [inline packages/alfa-predicate/test/predicate.spec.ts:1]: Looks like these are all tests for refinement. Should be completely killed here (or replaced by Predicate test).

**Review by Jym77 — CHANGES_REQUESTED** [review]
Looks fine.

File: packages/alfa-refinement/test/refinement.spec.ts (hunk 1 above)
Comment (kasperisager) [reply packages/alfa-refinement/test/refinement.spec.ts:1]: Good catch, completely forgot about those! 🤦

File: packages/alfa-predicate/test/predicate.spec.ts (hunk 2 above)
Comment (kasperisager) [reply packages/alfa-predicate/test/predicate.spec.ts:1]: Ditto on https://github.com/Siteimprove/alfa/pull/415#discussion_r496573586.

**Review by Jym77 — APPROVED** [review]

**Pull request merged**